plugins/connection/grpc.py

```diff
@@ -0,0 +1,292 @@
+      - Be sure to fully understand the security implications of enabling this
+        option as it could create a security vulnerability by logging sensitive information in log file.
+    default: False
+    ini:
+      - section: persistent_connection
+        key: log_messages
+    env:
+      - name: ANSIBLE_PERSISTENT_LOG_MESSAGES
+    vars:
+      - name: ansible_persistent_log_messages
+  grpc_type:
+    type: str
+    description: grpc type
+    default: False
+    ini:
+      - section: grpc_connection
+        key: type
+    env:
+      - name: ANSIBLE_GRPC_CONNECTION_TYPE
+    vars:
+      - name: ansible_grpc_connection_type
+"""
+
+from ansible.errors import AnsibleConnectionFailure, AnsibleError
+from ansible.plugins.connection import NetworkConnectionBase
+from importlib import import_module
+
+try:
+    from grpc import ssl_channel_credentials, secure_channel, insecure_channel
+    from grpc.beta import implementations
+    HAS_GRPC = True
+except ImportError:
+    HAS_GRPC = False
+
+try:
+    from google import protobuf
+    HAS_PROTOBUF = True
+except ImportError:
+    HAS_PROTOBUF = False
+
+
+class Connection(NetworkConnectionBase):
+    """GRPC connections"""
+
+    transport = 'grpc'
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        # TODO: Need to add support to make grpc connection work with non-network target host.
+        # Currently this works only with network target host.
+        if self._network_os:
+            if not HAS_PROTOBUF:
+                raise AnsibleError(
+                    "protobuf is required to use the grpc connection type. Please run 'pip install protobuf'"
+                )
+            os_split = self._network_os.split(".")
+            grpc_type = os_split[0] + "." + os_split[1] + ".grpc"
+            cref = dict(zip(["corg", "cname", "plugin"], grpc_type.split(".")))
+            grpclib = "ansible_collections.{corg}.{cname}.plugins.sub_plugins.grpc.{plugin}".format(
+                **cref
+            )
+            grpccls = getattr(import_module(grpclib), "Grpc")
+            grpc_obj = grpccls(self)
+
+            if grpc_obj:
+                self._sub_plugin = {'type': 'grpc', 'name': grpc_type, 'obj': grpc_obj}
+                self.queue_message('log', 'loaded gRPC plugin for network_os %s' % self._network_os)
+                self.queue_message('log', 'network_os is set to %s' % self._network_os)
+            else:
+                raise AnsibleConnectionFailure('unable to load API plugin for network_os %s' % self._network_os)
+        else:
+            self._sub_plugin['type'] = 'external'
+            self.queue_message('warning', 'Unable to automatically determine gRPC implementation type.'
+                                          ' Please manually configure ansible_network_os value for this host')
```

Comment:
Can update the error message to mention `grpc_type` as well, `Please manually configure ansible_network_os value or grpc_type configuration variable for this host`